### PR TITLE
Document --filter

### PR DESCRIPTION
**Pull request by gsoltis**
- [x] Adds the CLI reference docs for `--filter`
 - [x] Updates the scoping page to use `--filter`

### CODE REVIEW

**Comment by gaspar09** [comment]
Are these ready to go?

**Comment by gsoltis** [comment]
@gaspar09 I think we're waiting for `1.2.0` release

File: docs/pages/docs/features/filtering-packages.mdx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+import Callout from "../../../components/callout";
+import HeartIcon from "@heroicons/react/solid/HeartIcon";
+
+# Filtering Packages
+
+Filtering packages allows you to target your tasks to only the packages you care about.
+Turborepo supports a `pnpm`-like `--filter` flag that allows you to select the packages
+that will act as "entry points" into your monorepo's package/task graph.
+You can match packages by name, directory, include dependents/dependencies, and by changes
```
Comment (jaredpalmer) [inline docs/pages/docs/features/filtering-packages.mdx:14]:
```suggestion
You can filter packages by name, directory, include dependents/dependencies, and by changes
```

File: docs/pages/docs/features/filtering-packages.mdx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+in git history.
+
+`turbo` will run each task against each matched packages, ensuring that any dependencies
+are run first, according to the pipeline specification in [`turbo.json`](/docs/reference/configuration#pipeline).
```
Comment (jaredpalmer) [inline docs/pages/docs/features/filtering-packages.mdx:18]:
```suggestion
are run first, according to the `pipeline` specification in [`turbo.json`](/docs/reference/configuration#pipeline).
```

File: docs/pages/docs/features/filtering-packages.mdx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+## Filter Syntax
+
+Filters specify combinations of package(s)/apps, directories, and git commits to act as entrypoints
+for execution.
+
+Multiple filters can be combined to select distinct sets of targets. Additionally, filters
+can also exclude targets. A target that matches any filter and is not explicitly excluded will
+be in the final entrypoint selection.
+
+Filter syntax is based on `pnpm`'s filter syntax.
```
Comment (jaredpalmer) [inline docs/pages/docs/features/filtering-packages.mdx:29]:
```suggestion
Turborepo's filter syntax is based on pnpm's filter syntax, so it should feel familiar to pnpm users. 
```

File: docs/pages/docs/features/filtering-packages.mdx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+### Package specifier - Glob match package names
```
Comment (jaredpalmer) [inline docs/pages/docs/features/filtering-packages.mdx:31]:
```suggestion
### Filter by package
```

File: docs/pages/docs/features/filtering-packages.mdx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+Scoped packages (`@foo/bar`) can be abbreviated to just the package name (`bar`) as long as the package name
+is unique within the monorepo (e.g. `@foo/bar` exists, but `@baz/bar` does not).
+
+```sh
+turbo run build --filter=my-pkg
+turbo run build --filter=@foo/bar
+turbo run build --filter=admin-*
+```
+
+### Dependents specifier - Include dependents of matched packages
```
Comment (jaredpalmer) [inline docs/pages/docs/features/filtering-packages.mdx:43]:
```suggestion
### Include dependents of matched packages
```

File: docs/pages/docs/features/filtering-packages.mdx [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+Prepend `...` to the filter. If `a` depends on `b`, `...b` will select `a` and `b`.
+Optionally including a `^` (`...^b`) will select all of `b`'s dependents, but not `b` itself. In this case, just `a` will be
+selected.
+
+```sh
+turbo run build --filter=...b
+turbo run build --filter=...^b
+```
+
+### Dependencies specifier - Include dependencies of matched packages
```
Comment (jaredpalmer) [inline docs/pages/docs/features/filtering-packages.mdx:54]:
```suggestion
### Include dependencies of matched packages
```

File: docs/pages/docs/features/filtering-packages.mdx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+Append `...` to the end of the filter. If `a` depends on `b`, `a...` will select `a` and `b`.
+Optionally including a `^` (`a^...`) will select all of `a`'s dependencies, but not `a` itself. In this case, just `b` will be selected.
+
+```sh
+turbo run build --filter=a...
+turbo run build --filter=a^...
+```
+
+### Directory specifier - Match a directory or directory tree
```
Comment (jaredpalmer) [inline docs/pages/docs/features/filtering-packages.mdx:64]:
```suggestion
### Filter by directory or directory tree
```

File: docs/pages/docs/features/filtering-packages.mdx [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+Supports exact matches (`--filter=./apps/docs`) and globs (`--filter=./apps/*`).
+When combined with other components, must be enclosed in `{}` (`--filter=...{./libs/*}`).
+
+```sh
+turbo run build --filter=./apps/*
+turbo run build --filter=...{./libs/*}
+```
+
+### Commit specifier - Match changed packages
```
Comment (jaredpalmer) [inline docs/pages/docs/features/filtering-packages.mdx:74]:
```suggestion
### Filter by changed packages
```

File: docs/pages/docs/features/filtering-packages.mdx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+against the changed packages. For instance, to select `foo` if any of `foo`'s dependencies have changed in the last commit,
+you can pass `--filter=foo...[HEAD^1]`. Note that this feature is different from `pnpm`'s syntax.
+
+```sh
+turbo run build --filter=[HEAD^1]
+turbo run build --filter=...[origin/my-feature]
+turbo run build --filter=@foo/bar...[HEAD^1]
+```
+
+### Exclusion specifier - Exclude matched packages
```
Comment (jaredpalmer) [inline docs/pages/docs/features/filtering-packages.mdx:92]:
```suggestion
### Excluding packages
```

File: docs/pages/docs/features/meta.json [hunk 10]
```diff
@@ -1,6 +1,6 @@
 {
   "pipelines": "Pipelines",
-  "scopes": "Scoped Tasks",
```
Comment (jaredpalmer) [inline docs/pages/docs/features/meta.json:3]: **addition:** Need to add 301 redirect in `next.config.js`

File: docs/pages/docs/features/filtering-packages.mdx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+- Test everything that has changed since the previous commit: `turbo run test --filter=[HEAD^1]`
+- Build the docs app, letting `turbo` infer the need to build dependencies from the pipeline: `turbo run build --filter=docs`
+- Test everything that depends on a library (`@scope/foo`), but not the library itself: `turbo run test --filter=...^@scope/foo`
+- Test each package in the `@scope` scope and in the `packages` directory, if it has changed since the previous commit:
+  `turbo run test --filter=@scope/*{packages/*}[HEAD^1]`
+
+<Callout type="idea" icon={<HeartIcon className="h-5 w-5 text-gray-400" aria-hidden="true" />}>
+Turborepo's Filter API design and docs were/are inspired by 
+[`pnpm`](https://pnpm.io/filtering)
```
Comment (jaredpalmer) [inline docs/pages/docs/features/filtering-packages.mdx:112]:
```suggestion
[pnpm](https://pnpm.io/filtering)
```

File: docs/pages/docs/reference/command-line-reference.mdx [hunk 12]
```diff
@@ -100,6 +100,25 @@ Task details include:
...
+`type: string[]`
+
+Specify combinations of package(s)/apps, directories, and git commits to act as entrypoints
+for execution.
+
+Multiple filters can be combined to select distinct sets of targets. Additionally, filters
+can also exclude targets. A target that matches any filter and is not explicitly excluded will
+be in the final entrypoint selection.
+
+[Filter syntax](/docs/features/filtering-packages) is based on `pnpm`'s filter syntax.
```
Comment (jaredpalmer) [inline docs/pages/docs/reference/command-line-reference.mdx:114]:
```suggestion
For more detailed information about the `--filter` flag and filtering, refer to the [dedicated page in our documentation](/docs/features/filtering-packages)
```

**Review by jaredpalmer — CHANGES_REQUESTED** [review]

File: docs/next.config.js [hunk 13]
```diff
@@ -32,6 +32,11 @@ module.exports = withNextra({
         permanent: true,
         destination: "/docs/guides/monorepo-tools",
       },
+      {
+        source: "/docs/features/scopes",
+        permanent: true,
+        destination: "/docs/features/filtering-packages",
```
Comment (jaredpalmer) [inline docs/next.config.js:38]: Let's change url to just `filtering`

**Review by jaredpalmer — CHANGES_REQUESTED** [review]

**Comment by jaredpalmer** [comment]
**suggestion/addition:** let's add a `<Callout>` about how legacy `--scope` is going to be deprecated, but is equivalent to `--filter=XXXXXX` for compatibility as of `turbo@1.1.2+`

**Comment by gsoltis** [comment]
Re: deprecation, `--scope`, `--since`, `--no-deps`, and `--include-dependencies` (but not `--ignore`) are all technically deprecated.

**Comment by jaredpalmer** [comment]
Then we should make a little "Deprecated in `1.2.x`" red bubble or `<Deprecated />` like we have for `<Beta>` in Remote Caching and add it to all of the deprecated flags

File: docs/pages/docs/features/filtering.mdx [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+Prepend `!` to the filter. Matched packages from the entire filter will be excluded from the set of targets.
+For example, match everything except `a`: `--filter=!a`. Note that you may need to escape `!` as appropriate for your shell (e.g. `\!`).
+
+```sh
+turbo run build --filter=!foo
+turbo run build --filter=./apps/* --filter=!./apps/admin
+```
+
+### Examples
```
Comment (megbird) [inline docs/pages/docs/features/filtering.mdx:102]: It seems like these examples might be better divided up in the sections that are the most relevant ☝️ just so the usage and examples are closer together.

**Review by megbird — COMMENTED** [review]
Added some non-blocking comments on the overall organization 😄

**Comment by weyert** [comment]
How would you rewrite this command to use `--filter` to avoid the deprecated arguments?
`turbo run build --cache-dir=.turbo-cache --include-dependencies --no-deps` or
`turbo run build --cache-dir=.turbo-cache --include-dependencies`

Maybe worth it to add these to the docs or in the blog post?  Also posted this question in the blog post PR

**Review by jaredpalmer — APPROVED** [review]

**Pull request merged**